### PR TITLE
fix Issue 22134 - Deprecate returning a discarded void value from a function

### PR DESCRIPTION
**Pull request by ibuclaw**
Rationale, `ReturnStatement` rewrites this as:
```
this.arr[i], return;
```
So it should undergo the same `discardsValue` checking as what `CommaExp` or `ExpStatement` does.

### CODE REVIEW

**Comment by ibuclaw** [comment]
FYI @RazvanN7 as this takes aim at code in libdparse.

File: src/dmd/statementsem.d
Comment (kinke) [inline src/dmd/statementsem.d:3313]: [could be `const`, but oh well]

**Review by kinke — APPROVED** [review]

**Comment by ibuclaw** [comment]
buildkite/phobos builds fail because deprecations are errors in that pipeline:
```
../build/.dub/packages/libdparse-0.15.4/libdparse/src/dparse/stack_buffer.d(76,19): Deprecation: `this.arr[i]` has no effect
../build/.dub/packages/libdparse-0.15.4/libdparse/src/dparse/parser.d(167,21): Error: template instance `dparse.parser.Parser.ownArray!(AliasInitializer)` error instantiating
```
libdparse will need to be fixed before this can be merged.

**Comment by Geod24** [comment]
Release notes please

**Comment by RazvanN7** [comment]
> libdparse will need to be fixed before this can be merged.

PR: https://github.com/dlang-community/libdparse/pull/446

**Comment by thewilsonator** [comment]
ping @ibuclaw

**Comment by RazvanN7** [comment]
@thewilsonator Libdparse has been fixed. Unfortunately, updating dscanner with the latest version of libdparse has unveiled a bug in dscanner [1] or dsymbol [2] that makes it impossible to update phobos [3] to use the latest version of dscanner. I have trying to contact someone that can fix this, but wasn't successful.

[1] https://github.com/dlang-community/D-Scanner/issues/827
[2] https://github.com/dlang-community/dsymbol/issues/157
[3] https://buildkite.com/dlang/phobos/builds/5801#b0c48543-18c3-4ccc-a242-705162085954

File: src/dmd/statementsem.d
Comment (ljmf00) [inline src/dmd/statementsem.d:3317]: Is this empty if really necessary?

Comment (ibuclaw) [reply src/dmd/statementsem.d:3317]: Uncomment when deprecation is over to turn it into an error.

Comment (maxhaton) [reply src/dmd/statementsem.d:3317]: This should be noted in the source code

**Comment by RazvanN7** [comment]
@ibuclaw dscanner has been fixed and phobos has been updated to the latest version of it. could you please rebase this and see if it passes all the tests?

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@ibuclaw I rebased and added the missing bits. @maxhaton @ljmf00 please take a look.

**Comment by ljmf00** [comment]
> @ibuclaw I rebased and added the missing bits. @maxhaton @ljmf00 please take a look.

CI is failing on https://github.com/dlang/dmd/blob/master/test/fail_compilation/noreturn.d#L72 and buildkite fails on Phobos as mentioned by @ibuclaw . This seems to be failing on non-void values, such as `typeof(*null)`. There is an excuse I have about `cast()` which seems to be not documented in the specification `10.18.4. Cast Expressions` and therefore I'm not sure what is the behaviour here specifically.

**Comment by ibuclaw** [comment]
Deprecation look valid to me, and `@noreturn` code-gen has a bug in it.  Because the variable doesn't have any side effects, and really is discarded.
```
noreturn a;
return a;  // re-written as: `a, assert(false)`.
```

**Comment by MoonlightSentinel** [comment]
> ```D
> noreturn a;
> return a;  // re-written as: `a;`, the return part is dropped.
> ```

This should actually be rewritten as `[return] assert(false);`, but that shouldn't block this PR

**Comment by ibuclaw** [comment]
Still blocked by libdparse not being updated.

> libdparse 0.15.4: building configuration "library"...
../build/.dub/packages/libdparse-0.15.4/libdparse/src/dparse/stack_buffer.d(76,19): Deprecation: `this.arr[i]` has no effect

**Comment by RazvanN7** [comment]
Rebased and fixed tests. Maybe this time we are lucky with libdparse.

**Comment by ljmf00** [comment]
I don't follow why this is marked as `Blocked`. @RazvanN7 can you specify?

**Comment by ibuclaw** [comment]
> I don't follow why this is marked as `Blocked`. @RazvanN7 can you specify?

Probably still blocked by the [noreturn bug](https://github.com/dlang/dmd/pull/12898#issuecomment-997299511).

**Comment by RazvanN7** [comment]
Indeed: https://issues.dlang.org/show_bug.cgi?id=23063

**Comment by RazvanN7** [comment]
Actually, there shouldn't be any deprecations triggered now since:

```d
{
     noreturn a;
     return a;
}
```

will be lowered to: 

```d
{
     noreturn a;
     assert(0);
}
```

**Comment by ibuclaw** [comment]
> Actually, there shouldn't be any deprecations triggered now since:
> 
> ```d
> {
>      noreturn a;
>      return a;
> }
> ```
> 
> will be lowered to:
> 
> ```d
> {
>      noreturn a;
>      assert(0);
> }
> ```

Exactly, I don't think I was the person who edited those tests though. :-)

**Comment by RazvanN7** [comment]
ah, closed it by mistake.

**Comment by RazvanN7** [comment]
> Exactly what I was saying, I don't think I was the person who edited those tests though. :-)

It seems I have modified those tests in a previous commit before actually fixing the noreturn bug.

File: src/dmd/statementsem.d
Comment (ibuclaw) [inline src/dmd/statementsem.d:2917]:
```suggestion
                // @@@DEPRECATED_2.111@@@
```

File: src/dmd/statementsem.d
Comment (ibuclaw) [reply src/dmd/statementsem.d:2917]:
```suggestion
                const olderrors = global.startGagging();
```

**Pull request merged**
